### ISSUE TITLE
Accept header too narray for external YAML references

### ISSUE THREAD

**Issue by ASchmidt1** [swagger-api/swagger-editor#1558]
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  Bug
| Which Swagger/OpenAPI version?      | 3.0.0
| Which Swagger-Editor version?       | editor.swagger.io
| How did you install Swagger-Editor? | using hosted version
| Which broswer & version?            | Firefox 56.0.2
| Which operating system?             | Windows 10

### Demonstration API definition
```yaml
paths:
  /superiors/{class}/{uid}:
    $ref: "http://staging.anamera.com/anamera_v2/schema/Path_Superiors.yaml"
# NOTE: the Editor itself must be running with a non-SSL URL to reference a non-SSL external URL.
```

### Expected Behavior
Content of YAML document should be inserted.

### Current Behavior
Error: "Could not resolve reference because of: Not Acceptable"

### Possible Solution
According to https://swagger.io/specification/#format-11 "...a JSON object, which may be represented either in JSON or YAML format." Consequently, the editor should create ACCEPT headers that will permit either JSON or YAML content types to be served - instead, the editor generates a highly restrictive Accept header:
 Origin: http://editor.swagger.io 
 Accept-Language: en-US,en;q=0.5
 Accept: application/json

The media type registration for "application/json" (https://datatracker.ietf.org/doc/draft-ietf-jsonbis-rfc7159bis/?include_text=1) is EXCLUSIVELY for JSON formatted content. It does NOT permit YAML content. Therefore, it's self-evident, that the "Accept" header generated by the swagger.io editor would not permit serving YAML formatted content.

For proper function in compliance with the OpenAPI specifications, the editor should use an Accept header similar to this:
Accept: application/json, application/x-yaml, text/vnd.yaml
(a minimum both the commonly used custom type since is registered yet, as well as the VND type that was proposed in 2015), or more liberally:
Accept: application/json, application/x-yaml, text/vnd.yaml, text/yaml, \*.\*;0.8

### Context
At present the only way for a server to respond with YAML formatted data, is to either IGNORE the non-matching ACCEPT request header entirely (if that's an option), or if a server is strict, one has to mislable YAML formatted data by storing it in a .json file type, so that it's served with the factually incorrect (and non-compliant) "application/json" Content-Type response header.

**Comment by webron** [comment]
This is indeed an issue that needs to be addressed, but it's also a configuration option in the server. The `Accept` header is what the client _requests_, but the server does not have to comply with that request if it is not familiar with the media type, and can reply with whatever it wants (providing the proper media type in the `content-type` header, of course). There is no technical limitation of having the client ask for JSON and the server replying with YAML.

**Comment by ASchmidt1** [comment]
Thank you webron for your comment. 
As I had stated under "Context" I am already aware of the available, temporary work-arounds (such as changing the server configuration to be lenient). 
The motiviation of documenting bugs is to eventually make some of the countless work-arounds obsolete which server administrators have to maintain (and remember) at any given time as the only interim way to deal with bugs in the myriads of technologies they are expected to support.

**Comment by webron** [comment]
Right, I'm only saying that this is not a workaround but a perfectly acceptable solution.

**Comment by ASchmidt1** [comment]
I suppose we just disagree on the definition or semantics of "solution":

Whenever you have to "work around" the fact that the underlying problem persists, by IGNORING THE EFFECTS of the problem - then obviously you have not solved the problem of the imprecise Accept header, you literally have implemented a "work-around".

(If our system intentionally enforces strict with Accept... headers we receive from various external sources besides "swagger.io", then being forced to "open up" our system to imperfect/non-compliant/vague outside requests is something we might grudgingly accept temporarily - but only until the original technical shortcoming is addressed.)

**Comment by webron** [comment]
You're right, we disagree. There's nothing imperfect/non-compliant/vague about that solution. I'd argue that we should just send `*/*` and let the server decide to return whatever it wants. That's not really different than having the server return whatever it wants regardless of the `Accept` header.

**Comment by ASchmidt1** [comment]
We fully agree - there would be NO problem if \*/\* was used. THAT would be a solution.

It would mean that the server is free to serve its default content-type. THAT is perfectly reasonable - the assumption being that the client's operator knows that they are doing and are willing to cope with whatever we'll deliver.

That is not at all the same than the client explicitly stating that it can only accept the wrong content type for the format requested. To which our server will respond that it has nothing to match that literal request. Again, our system assumes that the client's operator is competent and it is better for them to clarify their ambiguity.

**Comment by webron** [comment]
Again, we don't fully agree. The client doesn't state what it can only accept. It states what it _wants_ to accept. The server does not have to comply with that request. There's no ambiguity. The server is not saying "hey, I'm telling you this is a JSON but it's in fact YAML", it's saying - "This is what I can give you, it's YAML, hope you can work with it.". Nothing ambiguous, nothing wrong with it. This does not go against the HTTP spec and it's a perfectly valid solution.

**Comment by ASchmidt1** [comment]
http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html

"Accept headers can be used to **indicate that the request is specifically limited** to a small set of desired types" ... "If an Accept header field is present, and if the server cannot send a response which is acceptable according to the combined Accept field value, then **the server SHOULD send a 406 (not acceptable) response.** "

and I know you do appreciate the unique semantic meaning that "SHOULD" has in the context of RFCs.

Therefore, our server fully complies with RFC2616 by responding with a "406" response, because it sending a content-type "application/json" with YAML formatted response would violate other RFCs that explicitly define the "application/json" content-type as being reserved for JSON formatted responses.

It is my believe that the Swagger editor should ENABLE the remote server to implement the application RFCs literally (while I realize that a reasonable work-around is available in the meantime.)

**Comment by shockey** [comment]
@ASchmidt1, we hashed this out in Swagger-UI recently: https://github.com/swagger-api/swagger-ui/issues/4070

We send `Accept: application/json, application/yaml` now, does that work for your purposes?

**Comment by ASchmidt1** [comment]
I very much do appreciate this now having been recognized as a bug, and a solution being proposed.

Allow me to comment that the media type chosen is NOT registered – nor am I aware that such a registration has even been submitted. 

Contrary what’s being implied IANA will NOT “get around to” arbitrarily assign a media type. It is up to the group managing the YAML specifications to decide upon - and then apply for - a media type to be assigned. I remember them making the argument that while “JSON” was intended to be primarily “machine readable” (thus the application/json type), one motivation for “YAML” was for it to be a easily human readable and editable document (thus the text/yaml type being more applicable).

I therefore question how factual/informed the statement “from what i can tell, application/yaml is likely to be the canonical media type for YAML going forward (once the IANA actually gets around to it...)” is. Can the author of that statement support it by citing the pending registration application, or at minimum the YAML specs, that points to “application/yaml” as the soon-to-come new type?

I respectfully question the wisdom of Swagger ignoring the specs and standardization processes, in favor of unilaterally/casually implementing a different, unregistered media type?

**Comment by shockey** [comment]
@ASchmidt1, there wasn't a lot of art behind choosing to add `application/yaml`: in the absence of any standards guidance on which is best, I chose what appears to be most widely supported by server implementations.

If you'd like to improve on what we have now, feel free to open a pull request in Swagger-Client. It's a simple one-line change to the Accept header, see here for an example: https://github.com/swagger-api/swagger-js/pull/1232/files#diff-166c7983bf6f3e2e63b5c6fe74b6c2d9L257

**Comment by Amachua** [comment]
Hello, I reoppen this thread since we're facing the same kind of issue that @ASchmidt1 had and I'm not sure to understand the reason that lead to this decision. Anyway, if I open a Pull Request with the change on the line you've suggested, would you validate it directly? :)

**Comment by webron** [comment]
What would the PR change?

**Comment by ASchmidt1** [comment]
I don't have the programming expertise, but my suggestion would be that the PR changes the Accept header to _minimally_ support the **text/vnd.yaml** MIME type that was explicitly stated/recommended in 2015. 

But given that no entity has ever actually registered a formal type for YAML, I would recommend a more liberal Accept header to wrap this matter up for good:

`Accept: application/json, application/x-yaml, text/vnd.yaml, text/yaml, *.*;0.8`

because this will:
a) cover the JSON type,
b) include the YAML type used in the field by convention,
c) include the YAML type explicitly mentioned in the 2015 proposal, and thus used in the field, since then,
d) includes a fall-back to accommodate all other variations that have sprung up due to the lack of any standardization.

This way, it will also be backward compatible in case anyone was using "application/yaml", even though the YAML format's stated purpose is for a human-readable/editable format - which means it would most properly be a "text" type not an "application" type.

**Comment by Amachua** [comment]
To quote you @webron (hope I do get this sentence correctly):

>I'd argue that we should just send `*/*` and let the server decide to return whatever it wants. 

Since the location of this configuration has changed since the last update, I would simply update the [file](https://github.com/swagger-api/swagger-js/blob/master/src/constants.js) with the value `*/*`instead of its current value `application/json, application/yaml`. 

Like this every consumer of your product wouldn't have to update their configuration to match this restrictive `Accept` header. Also you'll be able to close this discussion and please the community (or, at least, @ASchmidt1 and myself ;)).

Finally, if you do accept this change for Swagger Editor, I've some people on my organization that also complains about this on Swagger Hub. If we do the update here, can you take the opportunity to also change the content negotiation behavior on your other tools? :)

**Comment by MichakrawSB** [comment]
We’ve released Swagger Editor v5! 🎉 Check out the details here: [Inside the New Swagger Editor](https://swagger.io/blog/inside-the-new-swagger-editor/).
We’re closing old issues related to previous versions. If you think any of them are still relevant, please open a new issue – this helps us prioritize what matters most. Thanks for your activity! 🚀
